sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -420,51 +417,71 @@ as we can guarantee we only need at most two buffers for any call to write (two
+                    try {
+                        long fileSize = channel.size();
+
+                        // If the file is larger than 256MB chunk it and stage it as blocks.
+                        if (fileSize > 256 * Constants.MB) {
+                            return uploadBlocks(fileSize, finalParallelTransferOptions, headers, metadata, tier,
+                                accessConditions, channel);
+                        } else {
+                            // Otherwise we know it can be sent in a single request reducing network overhead.
+                            return uploadFullBlob(headers, metadata, tier, accessConditions, channel, fileSize);
+                        }
+                    } catch (IOException ex) {
+                        return Mono.error(new UncheckedIOException(ex));
+                    }
                 }, this::uploadFileCleanup);
         } catch (RuntimeException ex) {
             return monoError(logger, ex);
         }
     }
 
+    private Mono<Void> uploadBlocks(long fileSize, ParallelTransferOptions parallelTransferOptions,
+        BlobHttpHeaders headers, Map<String, String> metadata, AccessTier tier,
+        BlobRequestConditions accessConditions, AsynchronousFileChannel channel) {
+        ProgressReceiver progressReceiver = parallelTransferOptions.getProgressReceiver();
+
+        // See ProgressReporter for an explanation on why this lock is necessary and why we use AtomicLong.
+        AtomicLong totalProgress = new AtomicLong();
+        Lock progressLock = new ReentrantLock();
+
+        final SortedMap<Long, String> blockIds = new TreeMap<>();
+        return Flux.fromIterable(sliceFile(fileSize, parallelTransferOptions.getBlockSize()))
+            .flatMap(chunk -> {
+                String blockId = getBlockID();
+                blockIds.put(chunk.getOffset(), blockId);
+
+                Flux<ByteBuffer> progressData = ProgressReporter.addParallelProgressReporting(
+                    FluxUtil.readFile(channel, chunk.getOffset(), chunk.getCount()),
+                    progressReceiver, progressLock, totalProgress);
+
+                return getBlockBlobAsyncClient()
+                    .stageBlockWithResponse(blockId, progressData, chunk.getCount(), accessConditions.getLeaseId());
+            })
+            .then(Mono.defer(() -> getBlockBlobAsyncClient().commitBlockListWithResponse(
+                new ArrayList<>(blockIds.values()), headers, metadata, tier, accessConditions)))
+            .then()
+            .doOnTerminate(() -> uploadFileCleanup(channel));
+    }
+
+    private Mono<Void> uploadFullBlob(BlobHttpHeaders headers, Map<String, String> metadata, AccessTier tier,
+        BlobRequestConditions accessConditions, AsynchronousFileChannel channel, long fileSize) {
+        return getBlockBlobAsyncClient().uploadWithResponse(FluxUtil.readFile(channel),
+            fileSize, headers, metadata, tier, accessConditions)
+            .then();
```

Comment:
I think you need uploadFileCleanup here, too.

Reply:
This is being called from within the `Mono.using` and the `resourceCleanup` consumer is the equivalent of a final so this shouldn't need to do a cleanup here.

Reply:
In that case, is the doOnTerminate on line 475 in uploadBlocks unnecessary?

Reply:
I don't think it is needed either.

Reply:
Removed